subprojects/scala/src/main/java/org/gradle/api/plugins/scala/ScalaBasePlugin.java

```diff
@@ -129,20 +149,29 @@ public void execute(ScalaCompile scalaCompile) {
                 scalaCompile.setDescription("Compiles the " + scalaSourceSet.getScala() + ".");
                 scalaCompile.setSource(scalaSourceSet.getScala());
 
+                scalaCompile.getAnalysisMappingFile().set(project.getLayout().getBuildDirectory().file("tmp/scala/compilerAnalysis/" + scalaCompile.getName() + ".mapping"));
 
-                // cannot use convention mapping because the resulting object won't be serializable
                 // cannot compute at task execution time because we need association with source set
-                // TODO: Replace this with Providers
                 IncrementalCompileOptions incrementalOptions = scalaCompile.getScalaCompileOptions().getIncrementalOptions();
-                if (incrementalOptions.getAnalysisFile() == null) {
-                    String analysisFilePath = project.getBuildDir().getPath() + "/tmp/scala/compilerAnalysis/" + scalaCompile.getName() + ".analysis";
-                    incrementalOptions.setAnalysisFile(new File(analysisFilePath));
-                }
-
-                if (incrementalOptions.getPublishedCode() == null) {
-                    Jar jarTask = (Jar) project.getTasks().findByName(sourceSet.getJarTaskName());
-                    incrementalOptions.setPublishedCode(jarTask == null ? null : jarTask.getArchivePath());
+                incrementalOptions.getAnalysisFile().set(
+                    project.getLayout().getBuildDirectory().file("tmp/scala/compilerAnalysis/" + scalaCompile.getName() + ".analysis")
+                );
+                // TODO: Replace this with Providers in Jar/AbstractArchiveTask
+                final Jar jarTask = (Jar) project.getTasks().findByName(sourceSet.getJarTaskName());
```

Comment:
The `findByName` makes it a little weirder.  We can't get a provider for the task because sometimes the task doesn't exist.

We look for a `jar` and a `testJar` task (for compiling test code).

I couldn't find any real documentation about how the analysis file and classpath entries are lined up.  I would prefer if we did this a little differently without involving the jar at all (if that works with Scala incremental compilation).